### PR TITLE
Framework: experiment with schemas to validate persisted redux data

### PR DESCRIPTION
**Pull request by gwwar**
### The Problem

If we persist redux data, what happens when our tree changes? Our data shape is different right? Currently we do nothing, which is a little like this:
datashapes

But as a dev, lets say we run master, let the redux tree persist, and then switch to a branch that has some minor refactoring for an existing subtree.

We then see something like:
trace

Not good. A normal user can hit this case too, just by visiting our website, and the coming back 2 weeks later.

(Don't worry `persist-redux` is off in all envs!)

### Solution

Let's check to see if our data shapes match. If the persisted data doesn't match our schemas, we throw it out and rebuild it from scratch.

As a continuation from #2754, this is an example of how we might use schemas to validate persisted redux data. 

As a recap, in each reducer, `SERIALIZE` takes what's in memory and returns a simple JS object. Similarly in `DESERIALIZE` the reducer is given state that is a simple JS object. With this PR, we then validate if the data matches the shape that we expect. If it is valid, we further manipulate the JS object to match what the subtree expects in memory (this might be a no-op, or say instantiation of Immutable.js objects). If the data is invalid, we return the default initial state.

There is some developer overhead for maintaining schemas, but I find that it's a great way to understand what's in the tree (without spinning up Calypso), and it gives us a lot of flexibility in refactoring our tree, since if the shape doesn't match, we rebuild it from scratch.

The following uses state/sites/items as an example using json schema v4 and [tv4](https://github.com/geraintluff/tv4)

## Testing Instructions
* In config/development.json set "persist-redux" to true
* Navigate to "My Sites"
* Switch between a few sites
* Reload the page
* Sites should be loaded without errors
* Tests in client/state/sites should also pass

### CODE REVIEW

**Comment by gwwar** [comment]
cc @mtias @aduth @blowery @rralian @dmsnell

File: client/state/sites/reducer.js [hunk 1]
```diff
@@ -27,15 +29,18 @@ export function items( state = {}, action ) {
 				[ action.site.ID ]: action.site
 			} );
 		case SERIALIZE:
-			// scrub _events, _maxListeners, and other misc functions
+			//TODO: site objects are being decorated in SitesList in lib/sites/sites-list/index.js
+			//with either lib/site/index.js or lib/site/jetpack.js
```
Comment (gwwar) [inline client/state/sites/reducer.js]: This is somewhat of a tangent, but I also checked what it would take to clean up sites in the redux tree. I ran into a few issues:
1) How do we cleanly pass the redux store to SitesList, where the api requests are occurring?
2) We expect our in memory sites to be class instances of lib/sites/index.js or lib/site/jetpack.js. Should we move away from this pattern?

File: client/state/sites/test/reducer.js [hunk 2]
```diff
@@ -94,14 +103,14 @@ describe( 'reducer', () => {
 					}
 				} );
 			} );
-			it.skip( 'when state has validation errors on DESERIALIZE it returns initial state', () => {
+			it( 'when state has validation errors on DESERIALIZE it returns initial state', () => {
```
Comment (aduth) [inline client/state/sites/test/reducer.js]: Just a minor vocabulary remark, conventionally `it` is provided an argument that causes the line to flow as human-readable text, i.e.:

`it( 'returns initial state when state has validation errors on DESERIALIZE', () => {`

File: client/state/sites/schema.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+const icon = {
+	type: 'object',
+	properties: {
+		img: { type: 'string' },
+		ico: { type: 'string' }
+	}
+};
+
+const logo = {
```
Comment (aduth) [inline client/state/sites/schema.js]: Personal preference here, but I think splitting with such granularity hurts readability more than it helps. If it were me, I'd limit it to `site` object at most, if not simply exporting the entire schema object:

```js
export default {
	type: 'object',
	patternProperties: {
		'^\\d+$': {
			type: 'object',
			properties: {
				ID: { type: 'number' },
				name: { type: 'string' },
				description: { type: 'string' },
				URL: { type: 'string' },
				jetpack: { type: 'boolean' },
				post_count: { type: 'number' },
				subscribers_count: { type: 'number' },
				lang: { type: 'string' },
				icon: {
					type: 'object',
					properties: {
						img: { type: 'string' },
						ico: { type: 'string' }
					}
				},
				logo: {
					type: 'object',
					properties: {
						id: { type: 'number' },
						sizes: { type: 'array' },
						url: { type: 'string' }
					}
				},
				visible: { type: 'boolean' },
				is_private: { type: 'boolean' },
				is_following: { type: 'boolean' },
				options: { type: 'object' },
				meta: { type: 'object' },
				user_can_manager: { type: 'boolean' },
				is_vip: { type: 'boolean' },
				is_multisite: { type: 'boolean' },
				capabilities: {
					type: 'object',
					patternProperties: {
						'^[a-z_]+$': { type: 'boolean' }
					}
				},
				plan: {
					type: 'object',
					properties: {
						product_id: { type: 'number' },
						product_slug: { type: 'string' },
						product_name_short: { type: 'string' },
						free_trial: { type: 'boolean' }
					}
				},
				single_user_site: { type: 'boolean' },
				domain: { type: 'string' },
				slug: { type: 'string' },
				title: { type: 'string' }
			},
			required: [ 'ID', 'name' ]
		}
	},
	additionalProperties: false
};
```

File: client/state/utils.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ */
+import config from 'config';
+
+/**
+ * Module variables
+ */
+
+export function isValidStateWithSchema( state, schema, checkForCycles = false, banUnknownProperties = false ) {
+	const result = tv4.validateResult( state, schema, checkForCycles, banUnknownProperties );
+	if ( ! result.valid && config( 'env' ) === 'development' ) {
```
Comment (aduth) [inline client/state/utils.js]: Consider using [`lib/warn`](https://github.com/Automattic/wp-calypso/blob/master/client/lib/warn/index.js) instead, as it encapsulates this logic already.

File: package.json [hunk 5]
```diff
@@ -95,6 +95,7 @@
     "superagent": "1.2.0",
     "tinymce": "4.2.8",
     "to-title-case": "0.1.5",
+    "tv4": "1.2.7",
```
Comment (aduth) [inline package.json]: While the package is released under public domain, meaning no attribution should be required ([IANAL](https://en.wikipedia.org/wiki/IANAL)), we might consider including this in `CREDITS.md`.

**Comment by aduth** [comment]
I like the choice of `tv4` for schema validation. It's small (~8kb gzipped), has [some popularity](https://github.com/geraintluff/tv4/stargazers), and implements a [well-defined spec](http://json-schema.org/). If anything, there's a notable lack of [recent updates](https://github.com/geraintluff/tv4/commits/master), though perhaps this means it's reasonably stable enough to not warrant frequent changes.

File: client/state/sites/reducer.js (hunk 1 above)
Comment (aduth) [inline client/state/sites/reducer.js]: >I also checked what it would take to clean up sites in the redux tree.

It would be nice if we could somehow intercept the received site prior to it being modified in the `sites-list` module. Perhaps [initializing the global instance](https://github.com/Automattic/wp-calypso/blob/master/client/lib/sites-list/index.js) early in the page lifecycle, and having the list module be aware of the Redux store, dispatching the received site upon being received? This seems admittedly awful as a long-term solution, but the long-term solution is to remove the `sites-list` altogether anyways.

Otherwise, maybe we should consider performing the site object cleanup from the `SITE_RECEIVE` reducer handler, so that pages/components making use of sites Redux state don't come to expect these properties and functions to continue to exist.

>Should we move away from this pattern?

Yes. If context-specific behavior is desired, these could probably be reimplemented in the selectors and actions, e.g.

```jsx
if ( isJetpackSite( state, siteId ) ) {
	dispatch( requestAvailableJetpackUpdates( siteId ) );
}
```

File: client/state/utils.js (hunk 4 above)
Comment (gwwar) [inline client/state/utils.js]: The only problem I ran into was that it was somewhat difficult to mock out the console.warn.bind in test. The only work around I found so far was using mockery and using a require on the reducer:
```
	before( function() {
		mockery.registerMock( 'lib/warn', () => {} );
		mockery.enable();
		items = require( '../reducer' ).items;
	} );
	after( function() {
		mockery.deregisterAll();
		mockery.disable();
	} );
```

Comment (aduth) [inline client/state/utils.js]: >The only problem I ran into was that it was somewhat difficult to mock out the console.warn.bind in test.

Yeah, I might imagine that to be the case. Generally, I think we shouldn't let tests dictate how best to write the function, and the mocking is only marginally more verbose.

File: package.json (hunk 5 above)
Comment (gwwar) [inline package.json]: added tv4 and localforage licenses to CREDITS.md in bc58d49

File: client/state/sites/test/reducer.js (hunk 2 above)
Comment (gwwar) [inline client/state/sites/test/reducer.js]: Good catch!

File: client/state/utils.js (hunk 4 above)
Comment (gwwar) [inline client/state/utils.js]: I went ahead with the mockery setup for now. I might play around with sinon sandbox later so we can have something simpler.

File: client/state/sites/reducer.js (hunk 1 above)
Comment (gwwar) [inline client/state/sites/reducer.js]: > Otherwise, maybe we should consider performing the site object cleanup from the SITE_RECEIVE reducer handler

This sounds good as a first step, and move where SITE_RECEIVE is being fired for #2757 Some of the sites-list lifecycle changes might be tricky and would be better in a separate PR.

**Comment by gwwar** [comment]
@aduth Thanks for the :eyes: ! Great feedback.

File: client/state/sites/reducer.js [hunk 6]
```diff
@@ -23,19 +24,20 @@ import { SITE_RECEIVE, SERIALIZE, DESERIALIZE } from 'state/action-types';
 export function items( state = {}, action ) {
 	switch ( action.type ) {
 		case SITE_RECEIVE:
+			//TODO: do not pass a decorated site object to SITE_RECEIVE
+			//site objects are being decorated in SitesList in lib/sites/sites-list/index.js
+			//with either lib/site/index.js or lib/site/jetpack.js
+			const blackList = [ '_headers', 'fetchingSettings', 'fetchingUsers',
```
Comment (blowery) [inline client/state/sites/reducer.js]: how nasty would it be to use a whitelist? tends to be safer and more future proof.

File: client/state/sites/test/reducer.js [hunk 7]
```diff
@@ -71,7 +81,7 @@ describe( 'reducer', () => {
 					2916284: { ID: 2916284, name: 'WordPress.com Example Blog' }
 				} );
 			} );
-			it( 'it loads state on DESERIALIZE', () => {
+			it( 'validates state on DESERIALIZE', () => {
 				const original = Object.freeze( {
```
Comment (blowery) [inline client/state/sites/test/reducer.js]: might want to deepfreeze this?

File: client/state/utils.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ * Internal dependencies
+ */
+import warn from 'lib/warn';
+
+/**
+ * Module variables
+ */
+
+export function isValidStateWithSchema( state, schema, checkForCycles = false, banUnknownProperties = false ) {
+	const result = tv4.validateResult( state, schema, checkForCycles, banUnknownProperties );
```
Comment (blowery) [inline client/state/utils.js]: how does banUnknown defaulting to false play with `additionalProperties: false` in the actual schema? who wins?

File: client/state/sites/reducer.js (hunk 6 above)
Comment (gwwar) [inline client/state/sites/reducer.js]: I was hoping to fix this properly soon by passing the resource before it's decorated to fix #2757

**Comment by blowery** [comment]
tests pass, trying the code

File: client/state/utils.js (hunk 8 above)
Comment (gwwar) [inline client/state/utils.js]: banUnknown wins, this is generally only useful if you're debugging something. I added checkForCycles, and bandUnknownProperties to the interface so I wouldn't have to look up tv4 documentation later for the extra bools. I can remove this if its distracting.

**Comment by blowery** [comment]
code works. my comments are just minor things, feel free to :ship:

File: client/state/sites/test/reducer.js (hunk 7 above)
Comment (gwwar) [inline client/state/sites/test/reducer.js]: :+1: :snowman: added in e29ce8c

**Comment by gwwar** [comment]
Thanks for looking @blowery and @aduth! I'm going to squish commits, and follow up this PR with more documentation on why we need this, and how to implement these schemas.
